// Copyright 2020 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package xls.synthesis;

import "xls/codegen/module_signature.proto";

message CompileRequest {
  optional string module_text = 1;
  optional xls.verilog.ModuleSignatureProto signature = 2;
  optional string top_module_name = 3;
  optional int64 target_frequency_hz = 4;
}

// TODO(leary): Hierarchical area report, cell count histogram report.
message InstanceCount {
  optional int64 combinational = 1;
  optional int64 sequential = 2;
  optional int64 buf_inv = 3;
  optional int64 total = 4;
  optional int64 latches = 5;
  optional int64 nets = 7;
  map<string, int64> cell_histogram = 8;
}

message Path {
  optional string input_port_name = 1;
  optional int64 input_bit = 2;
  optional string output_port_name = 3;
  optional string output_bit = 4;
}

// Response to a CompileRequest.
message CompileResponse {
  optional int64 elapsed_runtime_ms = 1;
  optional int64 slack_ps = 2;
  optional float power = 3;
  // `area` is the total area of the design, which includes sequential area.
  // `area` does not include the cell types with unknown area.
  optional double area = 4;
  optional string netlist = 5;
  optional InstanceCount instance_count = 6;
  repeated Path failing_paths = 7;
  map<string, string> data_fields = 8;
  optional int64 max_frequency_hz = 9;
  optional string place_and_route_result = 10;
  // This tells the timing client that max_frequency_hz
  //  doesn't depend on the requested target frequency,
  //  which is true for Yosys with the current script.
  optional bool insensitive_to_target_freq = 11;
  optional double sequential_area = 12;
  optional int64 num_unknown_area_cell_types = 13;
}

// Encapsulates a series of compile results of a verilog module at various
// frequencies to determine the maximum frequency of the design.
message SynthesisSweepResult {
  // Verilog module text.
  optional string module_text = 1;

  // The signature of the Verilog module.
  optional xls.verilog.ModuleSignatureProto signature = 2;

  // Name of the top-level Verilog module.
  optional string top_module_name = 3;

  // The maximum frequency for which the design met timing. 0 if no such
  // frequency was found.
  optional int64 max_frequency_hz = 4;

  message SynthesisResult {
    optional int64 target_frequency_hz = 1;
    optional CompileResponse response = 2;
  }

  // The compile results of the various target frequencies attempted.
  repeated SynthesisResult results = 5;
}
